packages/firebase_auth/firebase_auth_platform_interface/lib/src/method_channel_firebase_auth.dart

```diff
@@ -0,0 +1,459 @@
+      case 'onAuthStateChanged':
+        _onAuthStageChangedHandler(call);
+        break;
+      case 'phoneVerificationCompleted':
+        final int handle = call.arguments['handle'];
+        final PhoneVerificationCompleted verificationCompleted =
+            _phoneAuthCallbacks[handle]['PhoneVerificationCompleted'];
+        verificationCompleted(PhoneAuthCredential(
+            jsonObject: call.arguments["phoneAuthCredential"].toString()));
+        break;
+      case 'phoneVerificationFailed':
+        final int handle = call.arguments['handle'];
+        final PhoneVerificationFailed verificationFailed =
+            _phoneAuthCallbacks[handle]['PhoneVerificationFailed'];
+        final Map<dynamic, dynamic> exception = call.arguments['exception'];
+        verificationFailed(
+            AuthException(exception['code'], exception['message']));
+        break;
+      case 'phoneCodeSent':
+        final int handle = call.arguments['handle'];
+        final String verificationId = call.arguments['verificationId'];
+        final int forceResendingToken = call.arguments['forceResendingToken'];
+
+        final PhoneCodeSent codeSent =
+            _phoneAuthCallbacks[handle]['PhoneCodeSent'];
+        if (forceResendingToken == null) {
+          codeSent(verificationId);
+        } else {
+          codeSent(verificationId, forceResendingToken);
+        }
+        break;
+      case 'phoneCodeAutoRetrievalTimeout':
+        final int handle = call.arguments['handle'];
+        final PhoneCodeAutoRetrievalTimeout codeAutoRetrievalTimeout =
+            _phoneAuthCallbacks[handle]['PhoneCodeAuthRetrievalTimeout'];
+        final String verificationId = call.arguments['verificationId'];
+        codeAutoRetrievalTimeout(verificationId);
+        break;
+    }
+  }
+
+  void _onAuthStageChangedHandler(MethodCall call) {
+    final Map<dynamic, dynamic> data = call.arguments['user'];
+    final int id = call.arguments['id'];
+
+    final PlatformUser currentUser = data != null ? _decodeUser(data) : null;
+    _authStateChangedControllers[id].add(currentUser);
+  }
+}
+
+PlatformUser _decodeUser(Map<dynamic, dynamic> data) {
+  final List<dynamic> rawProviderData = data['providerData'];
+  final List<Map<dynamic, dynamic>> castProviderData =
+      rawProviderData.cast<Map<dynamic, dynamic>>();
+  final List<PlatformUserInfo> providerData =
+      castProviderData.map<PlatformUserInfo>(_decodeUserInfo).toList();
+  return PlatformUser(
+    providerId: data['providerId'],
+    uid: data['uid'],
+    displayName: data['displayName'],
+    photoUrl: data['photoUrl'],
+    email: data['email'],
+    phoneNumber: data['phoneNumber'],
+    isAnonymous: data['isAnonymous'],
+    isEmailVerified: data['isEmailVerified'],
+    creationTimestamp: data['creationTimestamp'],
+    lastSignInTimestamp: data['lastSignInTimestamp'],
+    providerData: providerData,
+  );
+}
+
+PlatformUserInfo _decodeUserInfo(Map<dynamic, dynamic> data) {
```

Comment:
I think all these methods need `if (data == null) return null` at the beginning (or `assert(data != null)`, but I prefer the null return than throwing at runtime)